dodiscover/toporder/cam.py

```diff
@@ -0,0 +1,236 @@
+            Child of the parent node in the input `parent`,`child` edge
+        score_gains : NDArray
+            Matrix of the score gain.
+            score_gain[i,j] is the gain in score obtaied by addition of i -> j edge to the graph.
+        directed_paths : NDArray
+            Existing directed paths in the graph.
+        """
+        score_gains[parent, child] = self.inf  # do not select same edge twice
+        score_gains[child, parent] = self.inf
+        self._update_directed_paths(parent, child, directed_paths)
+        score_gains[np.transpose(directed_paths == 1, (1, 0))] = self.inf
+
+    def _initialize_score(self, X: NDArray, directed_paths: NDArray) -> Tuple[NDArray, NDArray]:
+        """
+        Initialize score gains matrix and the score contribution of each node.
+
+        If self.do_pns = True or self.do_pns is None and number of nodes >= 20.
+
+        Parameters
+        ----------
+        X : np.array
+            n x d matrix of the data
+        directed_paths : np.ndarray
+            d x d matrix encoding the directed paths in the graph.
+            directed_paths[i,j]=1 if there is a directed path from i to j.
+
+        Return
+        ------
+        score_gain : np.ndarray
+            d x d matrix of gains. score_gain[i, j] is the additive contribute to the score
+            (i.e. the gain) in adding i as parent of j
+        init_score : np.ndarray
+            d x 1 vector of the score contribute of each node.
+            Since the initial topological ordering is empty,
+            all nodes are initially treated as source.
+        """
+        _, d = X.shape
+        G_excluded = self.context.excluded_edges  # nx.Graph with excluded edges
+        if (self.do_pns) or (self.do_pns is None and d > 20):
+            A_pns = self.pns(A=np.ones((d, d)), X=X)
+            self.exclude_edges(A_pns)
+
+        # Initialize matrix of score gains and vector of initial scores
+        score_gains = np.zeros((d, d))
+        score_gains[
+            np.transpose(directed_paths == 1, (1, 0))
+        ] = self.inf  # avoid cycles setting entries to -Inf according to current directed_paths
+        nodes_variance = np.var(X, axis=0)
+        init_score = -np.log(nodes_variance)
+
+        # Score gain for each edge i -> j
+        for i in range(d):
+            for j in range(d):
+                if G_excluded.has_edge(i, j):
+                    score_gains[i, j] = self.inf
+                elif score_gains[i, j] != self.inf:
+                    gam = self._fit_model(X[:, i].reshape(-1, 1), X[:, j].reshape(-1, 1))
+                    residuals = X[:, j] - gam.predict(
+                        X[:, i].reshape(-1, 1)
+                    )  # Overfitting? Also, recomputing gam.predict(•) is likely inefficient
+                    gain = (
+                        -np.log(np.var(residuals)) - init_score[j]
+                    )  # np.var(residuals) different from 1/n * np.sum(np.square(residuals))...
```

Comment:
Is this a problem? Has it been addressed?

Reply:
- For comment 1, part 1 `# Overfitting?`: overfitting should be accounted for by the smoothing parameter lambda that is tuned in `_fit_model()` method. This is analogous to original R implementation of CAM
- For comment 1, part 2: `Also, recomputing gam.predict(•) is likely inefficient` I didn't find a way with `pygam` library to retain predictions after fitting, such that I have to recompute them. This is suboptimal, but the impact is relatively small as in `_fit_model()` we have to recompute predictions several times for the lambda tuning anyway. 
- Comment 2: `# np.var(residuals) different from 1/n * np.sum(np.square(residuals))` is addressed.